### PR TITLE
Extend HTTP client tracing by message body

### PR DESCRIPTION
**Pull request by DDvO**
* add `OSSL_TRACE_STRING`, `OSSL_TRACE_STRING_MAX`, and `OSSL_trace_string()`
* `OSSL_HTTP_REQ_CTX_nbio()`: use `OSSL_TRACE_STRING()` for msg body where it makes sense

### CODE REVIEW

**Comment by DDvO** [comment]
Ping for reviews

**Comment by wan7teen-ping** [comment]
Hi I will help you with everything you need but you have to subscribe to
the youtube channel and I will fulfill all your requests

https://youtu.be/aIaqr4eq18g

Pada Sab, 8 Okt 2022, 8:09 AM openssl-machine ***@***.***>
menulis:

File: crypto/trace.c [hunk 1]
```diff
@@ -519,3 +519,26 @@ void OSSL_trace_end(int category, BIO * channel)
...
+    if (!full && size > OSSL_TRACE_STRING_MAX) {
+        BIO_printf(out, "[len %zu limited to %d]: ",
+                   size, OSSL_TRACE_STRING_MAX);
+        len = OSSL_TRACE_STRING_MAX;
+    } else {
+        len = (int)size;
+    }
+    if (!text) { /* mask control characters */
+        for (i = 0; i < len; i++, data++)
+            buf[i] = (char)*data < ' ' ? ' ' : *data;
```
Comment (paulidale) [inline crypto/trace.c:538]: 127 is normally considered a control character too.
iscntrl() or ossl_iscntrl() perhaps?

**Review by paulidale — APPROVED** [review]

File: crypto/trace.c (hunk 1 above)
Comment (DDvO) [reply crypto/trace.c]: Good point - now using `ossl_iscntrl()`
and excempting the `\n` case in order to preserve existing newlines.

**Comment by DDvO** [comment]
Also rebased to fix trivial merge conflicts in `libcrypto.num` due to 3.1->3.2 changes

**Comment by DDvO** [comment]
Rebased to fix a new trivial merge conflict.

**Comment by DDvO** [comment]
Ping for 2nd review

File: doc/man1/openssl-cmp.pod.in [hunk 2]
```diff
@@ -1067,9 +1067,9 @@ although they usually contain hints that would be helpful for diagnostics.
 For assisting in such cases the CMP client offers a workaround via the
 B<-unprotected_errors> option, which allows accepting such negative messages.
 
-If OpenSSL was built with trace support enabled
+If OpenSSL was built with trace support enabled (e.g., C<./config enable-trace>)
 and the environment variable B<OPENSSL_TRACE> includes B<HTTP>,
-the request and response headers of HTTP transfers are printed.
+the requests and response headers of transferred via HTTP are printed.
```
Comment (t8m) [inline doc/man1/openssl-cmp.pod.in:1072]: The `of` seems superfluous?

File: doc/man3/OSSL_trace_enabled.pod [hunk 3]
```diff
@@ -287,6 +309,9 @@ L<OSSL_trace_set_channel(3)>, L<OSSL_trace_set_callback(3)>
 
 The OpenSSL Tracing API was added in OpenSSL 3.0.
 
+OSSL_TRACE_STRING(), OSSL_TRACE_STRING_MAX, and OSSL_trace_string
+were added in OpensSL 3.1.
```
Comment (t8m) [inline doc/man3/OSSL_trace_enabled.pod:313]: `OpenSSL 3.2`

**Review by t8m — CHANGES_REQUESTED** [review]
Would it be possible to add some tests or is too big effort to add them?

File: doc/man1/openssl-cmp.pod.in (hunk 2 above)
Comment (DDvO) [reply doc/man1/openssl-cmp.pod.in]: Right, so removed it.

File: doc/man3/OSSL_trace_enabled.pod (hunk 3 above)
Comment (DDvO) [reply doc/man3/OSSL_trace_enabled.pod]: Updated.

**Comment by DDvO** [comment]
Good point to add tests - done.
Doing so, I found a mem management bug and doc omission, fixed in a separate PR.
BTW, renamed `90-test_traceapi.t` to `90-test_trace_api.t` for consistency.

I had to rebase in order to fix merge conflicts.

File: doc/man3/OSSL_trace_enabled.pod [hunk 4]
```diff
@@ -287,6 +310,9 @@ L<OSSL_trace_set_channel(3)>, L<OSSL_trace_set_callback(3)>
 
 The OpenSSL Tracing API was added in OpenSSL 3.0.
 
+OSSL_TRACE_STRING(), OSSL_TRACE_STRING_MAX, and OSSL_trace_string
+were added in OpensSL 3.2.
```
Comment (t8m) [inline doc/man3/OSSL_trace_enabled.pod:314]: Nit: OpenSSL

File: test/trace_api_test.c [hunk 5]
```diff
@@ -65,18 +65,45 @@ static int test_trace_categories(void)
...
+/* OSSL_STR80 must have length OSSL_TRACE_STRING_MAX */
+#define OSSL_STR80 "1234567890123456789012345678901234567890123456789012345678901234567890123456789\n"
+#define OSSL_STR81 (OSSL_STR80"x")
+#define OSSL_CTRL "A\xfe\nB"
+#define OSSL_MASKED "A \nB"
+#define OSSL_BYE "Good Bye Universe\n"
+#define OSSL_END "-abc"
+
+#define trace_string(text, full, str) \
+    OSSL_trace_string(trc_out, text, full, (unsigned char*)(str), strlen(str))
```
Comment (t8m) [inline test/trace_api_test.c:80]: Nit: space between char and `*`

**Comment by DDvO** [comment]
Thanks for the immediate further comments.
Fixed both nits, as well as a mistake on indentation nesting of `#define`.

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
> CI is relevant.

Doc nit fixed.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@paulidale please reconfirm

**Review by paulidale — APPROVED** [review]

**Comment by DDvO** [comment]
Merged - thanks @t8m and @paulidale for your comments and approvals.
Glad that this is through now.

**Pull request closed without merging**